lib/mastodon/domains_cli.rb

```diff
@@ -74,18 +75,25 @@ def purge(domain = nil)
       default (`summary`), a summary of the statistics is returned. The other options
       are `domains`, which returns a newline-delimited list of all discovered peers,
       and `json`, which dumps all the aggregated data raw.
+
+      The --exclude-suspended (-x) option means that domains that are suspended
+      instance-wide do not appear in the output and are not included in summaries.
+      This also excludes subdomains of any of those domains.
     LONG_DESC
     def crawl(start = nil)
-      stats     = Concurrent::Hash.new
-      processed = Concurrent::AtomicFixnum.new(0)
-      failed    = Concurrent::AtomicFixnum.new(0)
-      start_at  = Time.now.to_f
-      seed      = start ? [start] : Account.remote.domains
+      stats           = Concurrent::Hash.new
+      processed       = Concurrent::AtomicFixnum.new(0)
+      failed          = Concurrent::AtomicFixnum.new(0)
+      start_at        = Time.now.to_f
+      seed            = start ? [start] : Account.remote.domains
+      blocked_domains = options[:exclude_suspended] ? DomainBlock.where(severity: 1).pluck(:domain) : []
 
       pool = Concurrent::ThreadPoolExecutor.new(min_threads: 0, max_threads: options[:concurrency], idletime: 10, auto_terminate: true, max_queue: 0)
 
       work_unit = ->(domain) do
         next if stats.key?(domain)
+        next if blocked_domains.any? { |blocked| domain.match(Regexp.new('\\.?' + blocked + '$')) }
```

Comment:
I suppose you could pre-compose a regex with all the domains, dunno which way is faster off the top of my head, it would be a big regex but you would save on an array iteration and initializing a new regex on each item.

Reply:
Compiling a giant regexp does seem significantly faster, I'll rewrite. 

```ruby
require 'Benchmark'

domains_to_test = []
blocked_domains = []
200000.times do blocked_domains.push(rand(400).to_s) end
15000.times do domains_to_test.push(rand(400).to_s) end

puts "Regexp.new"
puts Benchmark.measure {
  domains_to_test.each do |domain|
    blocked_domains.any? { |blocked| domain.match(Regexp.new('\\.?' + blocked + '$')) }
  end
}

puts "precompiled giant Regexp"
puts Benchmark.measure {
  reg = Regexp.new('\\.?' + blocked_domains.join('|') + '$')
  domains_to_test.each do |domain|
    domain.match(reg)
  end
}
```

yields

```
Regexp.new
  8.581110   0.078617   8.659727 (  8.678938)
precompiled giant Regexp
  0.231612   0.009794   0.241406 (  0.241989)
```